### ISSUE TITLE
MIDI import malfunctions

### ISSUE THREAD

**Issue by JordanKyser** [LMMS/lmms#4059]
When doing MIDI imports of more than 8 tracks, it only shows the first 8 tracks in the MIDI, plus the percussion track. When there are changes in volume, it shows all of the automations (as if all MIDI tracks are present).

An even stranger part to this bug is that the MIDI tracks after the 8th MIDI track have their patches placed on the first MIDI tracks.

I found this out as an LMMS bug because I tried it out on a different type of software (such as MuseScore (as a composition software) or FL) and they were imported normally.

**Comment by zonkmachine** [comment]
Can you add an assortment of the files you're testing with in a zip?

**Comment by JordanKyser** [comment]
http://www.mediafire.com/file/w4che5wi2fmhdrd/MIDI%20tests.zip

Okay, I have included 3 MIDI files and an additional MIDI file that shows a bug when you import a MIDI file with percussion instruments as separate MIDI tracks (they will give the right patch, but the wrong bank (in which the percussion bank is Bank 128, thus playing the default piano).

**Comment by musikBear** [comment]
I got a fake(?) positive warning for malware from that place, and an 'extension' to firefox, was deemed 'absolutely necessary' (ya right...)   :/ Mediafire is s*** - you can upload zip files directly here on hub, dont use mediafire. 
Here are the library scanned and direct uploaded here
[MIDI tests.zip](https://github.com/LMMS/lmms/files/1576474/MIDI.tests.zip)

@JordanKyser Believe your issue to be related to #2741 
and my comment here https://github.com/LMMS/lmms/issues/2741#issuecomment-352808884
I tried a midi0, and afaics, that changes the issue(?)
Here is the project
[Issue4059_midi_0_versionOfPecusError.txt](https://github.com/LMMS/lmms/files/1576510/Issue4059_midi_0_versionOfPecusError.txt)

**Comment by tresf** [comment]
> dont use mediafire [...] upload here

Agreed, but next time, just upload it and educate the user.  Mediafire is used for both good and bad software.  Let's keep the "malware" warnings out and just educate the users about the file attachment feature inside GitHub. 👍

**Comment by JordanKyser** [comment]
Oh okay, I will try that next time when I do a zip folder.

> Believe your issue to be related to #2741

I believe that this has to do with how the tracks are rendered and not how the notes are rendered, but I have had that problem as well.

I will try to send screenshots for this issue.

**Comment by JordanKyser** [comment]
Okay, I have the situations here.

----

SITUATION 1:

The first and main situation is how the final MIDI tracks are not shown after rendering, but you can still see their corresponding automation tracks.

HOW THE LAST MIDI TRACKS ARE RENDERED:
![Final MIDI tracks are missing](https://cdn.discordapp.com/attachments/372171748613029900/393132176394420224/Screenshot_76.png "Final MIDI tracks are missing")

HOW THE LAST MIDI TRACKS SHOULD BE RENDERED:
![Final MIDI tracks should be shown](https://cdn.discordapp.com/attachments/372171748613029900/393132187584561162/Screenshot_77.png "Final MIDI tracks should be shown")

----

SITUATION 2:

The removed MIDI tracks have their patches placed on the first MIDI tracks. In other words, if their are 7 MIDI tracks removed during the MIDI import, then the first MIDI tracks rendered will have the corresponding MIDI patches (as I have said earlier for the MIDI patches: 16 will go to 1, 17 will go to 2, 18 will go to 3, ... 22 will go to 7, 23 will go to 8, and so forth.

In this case, the first MIDI track that was removed will be a trombone, which will be MIDI track 16. The last MIDI track that was removed will be a Xylophone (known as the Mallets or Marimba), which will be MIDI track 23.

HOW THE FIRST MIDI TRACK IS RENDERED:
![First MIDI track is placed as the first MIDI track that was not rendered properly](https://cdn.discordapp.com/attachments/372171748613029900/393132160246087690/Screenshot_73.png "First MIDI track is placed as the first MIDI track that was not rendered properly")

HOW THE FIRST MIDI TRACK SHOULD BE RENDERED:
![First MIDI track should be what the track specifies, either specified in the track and/or the MIDI file itself](https://cdn.discordapp.com/attachments/372171748613029900/393132158845190144/Screenshot_72.png "First MIDI track should be what the track specifies, either specified in the track and/or the MIDI file itself")


HOW THE LAST MIDI TRACK IS RENDERED (in this case, MIDI track 8):
![First MIDI track is placed as the first MIDI track that was not rendered properly](https://cdn.discordapp.com/attachments/372171748613029900/393132166814498816/Screenshot_74.png "First MIDI track is placed as the first MIDI track that was not rendered properly")

HOW THE CORRESPONDING TRACK SHOULD BE RENDERED:
![First MIDI track should be what the track specifies, either specified in the track and/or the MIDI file itself](https://cdn.discordapp.com/attachments/372171748613029900/393132179699400734/Screenshot_75.png "First MIDI track should be what the track specifies, either specified in the track and/or the MIDI file itself")

**Comment by PhysSong** [comment]
https://github.com/LMMS/lmms/issues/815#issuecomment-45494509 seems related.

**Comment by JordanKyser** [comment]
> ...but looking a bit deeper, there's room to improve the MIDI import. As of now, it looks like LMMS splits a MIDI file into tracks by channel, which is a perfectly valid strategy, but maybe not the best for all files. There are three dimensions in play here:
- Tracks, a MIDI file can contain an "unlimited" number of them.
- Channels, a track can contain events for several channels and a channel can be used in more than one track.
- Programs, i.e. instrument sounds, which can change on a channel at any time, and there's only one program in effect on the channel at any given time.
There are use cases for all kinds of splits: The file at hand now would've been handled better by splitting on tracks, and importing a file with a lot of program changes to be used with some other plugin than the SF2 player could benefit from a split on channel and program. (program changes are now handled through automation on the patch and bank dials in the SF2 player, if my reading of the code is correct)

Yes, this is pretty much what is happening.
